src/Analyser/MutatingScope.php

```diff
@@ -5637,6 +5637,21 @@ private function methodCallReturnType(Type $typeWithMethod, string $methodName,
 		}
 
 		$methodReflection = $typeWithMethod->getMethod($methodName, $this);
+
+		// https://www.php.net/manual/en/language.oop5.late-static-bindings.php
+		if (
+			$methodCall instanceof Expr\StaticCall
+			&& $typeWithMethod instanceof StaticType
```

Comment:
This isn't a proper fix. The place where the StaticType transformation gets decided is in `ObjectType::getUnresolvedMethodPrototype()` and `StaticType::getUnresolvedMethodPrototype()`. One of them transforms `static` to the class it's called on, and the other one preserves `static`. Some piece of logic should be written there.

Reply:
But how - in `ObjectType::getUnresolvedMethodPrototype()` there is no info about how the call was made - ie. `$methodCall` and `$typeWithMethod` is not passed. Maybe you can look into the fix with the tests below?

Reply:
What kind of call do you want to transform and what kind of call do you not want to transform?

Reply:
See condition a few lines below - `['self', 'static', 'parent']` - even if `self` resolves to the same class where declared, it does not behave the same due php "late static binding".

Reply:
I've asked for examples of method calls.

Reply:
see the tests and/or https://github.com/phpstan/phpstan/issues/10469

Reply:
Yes, this comment explains it as I need it: https://github.com/phpstan/phpstan/issues/10469#issuecomment-1906046819

Reply:
Thanks - now the question is how to fix it.